nbqa/replace_source.py

```diff
@@ -87,24 +169,116 @@ def main(python_file: "Path", notebook: "Path", notebook_info: NotebookInfo) ->
-                    source,
-                    notebook_info.temporary_lines.get(code_cell_number, []),
-                )
-
+            new_source = _get_new_source(code_cell_number, notebook_info, pycells)
+            if new_source is not None:
+                cell["source"] = new_source
         new_cells.append(cell)
 
     notebook_json.update({"cells": new_cells})
     notebook.write_text(f"{json.dumps(notebook_json, indent=1, ensure_ascii=False)}\n")
+
+
+def print_red(message: str) -> str:
+    """
+    Print message in red.
+
+    Parameters
+    ----------
+    message
+        String to print in red.
+    """
+    return f"{RED}{message}{RESET}"
+
+
+def print_green(message: str) -> str:
+    """
+    Print message in green.
+
+    Parameters
+    ----------
+    message
+        String to print in green.
+    """
+    return f"{GREEN}{message}{RESET}"
+
+
+def _print_diff(code_cell_number: int, cell_diff: Iterator[str]) -> None:
+    """
+    Print diff between cells, colouring as appropriate.
+
+    Parameters
+    ----------
+    code_cell_number
+        Current cell number
+    cell_diff
+        Diff between original and new versions of cell.
+    """
+    peek, cell_diff = _peek(cell_diff)
+    if peek is not None:
+        header = f"Cell {code_cell_number}"
+        sys.stdout.write(f"{BOLD}{header}{RESET}\n")
+        sys.stdout.write(f"{'-'*len(header)}\n")
+        for line in cell_diff:
+            if line.startswith("+++") or line.startswith("---"):
+                sys.stdout.write(line)
+            elif line.startswith("+"):
+                sys.stdout.write(print_green(line))
+            elif line.startswith("-"):
+                sys.stdout.write(print_red(line))
+            else:
+                sys.stdout.write(line)
+        sys.stdout.write("\n")
+
+
+def add_newline_to_last_line(cell: List[str]) -> None:
+    """
+    Add trailing newline to last line in cell, as required by difflib.
+
+    Parameters
+    ----------
+    cell
+        Current cell being processed.
+    """
+    if cell:
+        cell[-1] += "\n"
+
+
+def diff(python_file: "Path", notebook: "Path", notebook_info: NotebookInfo) -> None:
+    """
+    Replace :code:`source` code cells of original notebook.
+
+    Parameters
+    ----------
+    python_file
+        Temporary Python file notebook was converted to.
+    notebook
+        Jupyter Notebook third-party tool is run against (unmodified).
+    notebook_info
+        Information about notebook cells used for processing
+    """
+    notebook_json = json.loads(notebook.read_text())
+
+    pycells = _get_pycells(python_file)
+
+    code_cell_number = 0
+    for cell in notebook_json["cells"]:
+        if cell["cell_type"] == "code":
+            code_cell_number += 1
+            new_source = _get_new_source(code_cell_number, notebook_info, pycells)
+            if new_source is not None:
+                add_newline_to_last_line(cell["source"])
```

Comment:
It would be better to check `if not cell["source"] and not new_source: `, so that the `unified_diff` function call can be simplified as 
```Python
cell_diff = unified_diff(
                    cell["source"].append("\n"),
                    new_source.append("\n"),
                    fromfile=str(notebook),
                    tofile=str(notebook),
                )
```

and `add_newline_to_last_line` function can be removed

Reply:
are you sure this would work? because `cell["source"].append("\n")` returns `None`, e.g.
```
>>> [1,2,3].append('\n') is None
True
```

Reply:
You are right. I changed `cell["source"] + ["\n"]` to use append but failed to move it out of the `unified_diff` function call.  You could replace `add_newline_to_last_line` with `cell["source"].append("\n")`. That was the intention.

Reply:
Adding one more line above with
```python
cell["source"].append("\n"), new_source.append("\n")
```
And just using `cell["source"]` and `new_source` in `unified_diff` should work though.

Reply:
@s-weigand are you sure? I still don't think it's the same - currently, I'm adding a newline to the last element of a list, rather than adding an extra element which is a newline.

i.e. if we start with
```
['foo\n', 'bar']
```
I want to end up with
```
['foo\n', 'bar\n']
```
rather than
```
['foo\n', 'bar', '\n']
```

Reply:
Now I understood. This `['foo\n', 'bar\n']` is how the source code looks inside `cell["source"]`. So what you did was the right way I feel, but still you don't need a method for it right?

Reply:
Probably not :smile: method removed

Reply:
@MarcoGorelli My comment was a comment to https://github.com/nbQA-dev/nbQA/pull/452#discussion_r523024054 . the gihub page didn't autorefresh 😝